### ISSUE TITLE
Queue not implemented for SwaV

### ISSUE THREAD

**Issue by ibro45** [lightly-ai/lightly#1006]
Hi, thanks for your great work!

From the [SwaV paper](https://arxiv.org/pdf/2006.09882.pdf):
<img width="1404" alt="image" src="https://user-images.githubusercontent.com/18015788/207389122-4adaffe6-eb36-481f-bcf6-92bbc2a7b11a.png">

[Lightly's `SwavLoss` ](https://github.com/lightly-ai/lightly/blob/7d3bc64ac3372c6e7ec8e24a8c56fb499209957f/lightly/loss/swav_loss.py#L65) does not implement this.

Here you can find [VISSL's implementation](https://github.com/facebookresearch/vissl/blob/346114aa1152aeadeb42d410951a95a28c53a207/vissl/losses/swav_loss.py#L191). It is quite messy though.

**Comment by guarin** [comment]
Hi, thanks a lot for pointing this out! You are right that we did not implement the queue.

There is sadly no easy way to add the queue to the loss as the features in the queue have to be matched to the prototypes which are stored in the model. We would have to store the queue outside of the loss. I think something like this could work (modified version of our [SwaV example](https://docs.lightly.ai/self-supervised-learning/examples/swav.html)):

```
class SwaV(nn.Module):
    def __init__(self, backbone):
        super().__init__()
        self.backbone = backbone
        self.projection_head = SwaVProjectionHead(512, 512, 128)
        self.prototypes = SwaVPrototypes(128, n_prototypes=512)

    def forward(self, x):
        x = self.backbone(x).flatten(start_dim=1)
        x = self.projection_head(x)
        x = nn.functional.normalize(x, dim=1, p=2)
        p = self.prototypes(x)
        return p, x    # return prototypes assignments and features

...

# Create one queue for each high resolution view.
queues = [MemoryBankModule(size=3, 840 ) for _ in range(2)]
for queue in queues:
    queue.to(device)

for epoch in range(10):
    total_loss = 0
    for batch, _, _ in dataloader:
        model.prototypes.normalize()
        multi_crop_outputs = [model(x.to(device)) for x in batch]
        multi_crop_prototypes = [prototypes for prototypes, _ in multi_crop_outputs]
        high_resolution = multi_crop_prototypes[:2]
        low_resolution = multi_crop_prototypes[2:]

        # Get queue features
        queue_prototypes = []
        with torch.no_grad():
            for queue, (_, projections) in zip(queues, multi_crop_outputs[:2]):
                queue_features = queue(projections, update=True)[1]
                queue_prototypes.append(model.prototypes(queue_features))

        loss = criterion(high_resolution, low_resolution, queue_prototypes)
        total_loss += loss.detach()
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
    avg_loss = total_loss / len(dataloader)
    print(f"epoch: {epoch:>02}, loss: {avg_loss:.5f}")
``` 

This would also require to add an optional `queue_outputs` argument to the forward method in the SwaV loss:
```
def forward(self, high_resolution_outputs, low_resolution_outputs, queue_outputs=None):
    n_crops = len(high_resolution_outputs) + len(low_resolution_outputs)

    # multi-crop iterations
    loss = 0.
    for i in range(len(high_resolution_outputs)):

        # compute codes of i-th high resolution crop
        with torch.no_grad():
            outputs = high_resolution_outputs[i].detach()

            # Append queue outputs
            if queue_outputs is not None:
                outputs = torch.cat((outputs, queue_outputs.detach()))

            q = sinkhorn(
                outputs,
                iterations=self.sinkhorn_iterations,
                epsilon=self.sinkhorn_epsilon,
                gather_distributed=self.sinkhorn_gather_distributed,
            )

            # Drop queue similarities
            if queue_outputs is not None:
                q = q[:len(high_resolution_outputs[i])]

        # compute subloss for each pair of crops
        subloss = 0.
        for v in range(len(high_resolution_outputs)):
            if v != i:
                subloss += self.subloss(high_resolution_outputs[v], q)

        for v in range(len(low_resolution_outputs)):
            subloss += self.subloss(low_resolution_outputs[v], q)

        loss += subloss / (n_crops - 1)

    return loss / len(high_resolution_outputs)
```


Not sure if this would work though, I didn't test it. It looks also a bit messy, maybe we can come up with a better design. I'll try to look at it in more detail once I have a bit more time.

I am also wondering if the missing queue could be the reason that SwaV performs so bad in our [Imagenette100 benchmarks](https://docs.lightly.ai/self-supervised-learning/getting_started/benchmarks.html#imagenet100).
